File: clients/typescript/src/transport/http.ts
import { ProtocolError } from "../errors.ts";
import { Input } from "./input.ts";
import { Timeout } from "../timeout.ts";
import { assert } from "../utils/common.ts";

export class RPC {
    static URL = new URL("/", "https://localhost:443");
    static TIMEOUT = Timeout.minute(2);

    static async call(id: number, input: Input, timeout: Timeout | null = RPC.TIMEOUT, url: URL = RPC.URL) {
        let headers: HeadersInit = {
            "content-type": "application/setu",
            "rpc-id": id.toString(),
        };

        if (timeout) {
            headers["rpc-timeout"] = timeout.toString();
        }
        let res = await fetch(url, { method: "POST", headers, body: input.channel.stream, signal: input.controller.signal });

        if (!res.ok) {
            throw new Error(`${res.statusText}: ${await res.text()}`);
        }

        let contentType = res.headers.get("content-type");

        assert(contentType == "application/setu", ProtocolError, () => `unexpected content-type: ${contentType ?? "none"}`);
        assert(res.body, ProtocolError, "No response body");

        return res.body;
    }
}

export interface Context {
    url?: URL,
    timeout?: Timeout | null
}

export function rpc(id: number, { timeout, url }: Context) {
    let input = new Input();
    let output = RPC.call(id, input, timeout, url);
    return [input, output] as const;
}

File: clients/dart/lib/transport/http.dart
import 'dart:typed_data';

import 'package:dio/dio.dart';

import 'package:setu_client/errors.dart';
import 'package:setu_client/http/client.dart';
import 'package:setu_client/setu/http_headers.dart';
import 'package:setu_client/timeout.dart';
import 'package:setu_client/utils/common.dart';

class RPC {
  static var uri = Uri.parse('https://localhost:443/');
  static var timeout = Timeout.minute(2);
  static var dio = createHttpClient();

  static Future<Stream<Uint8List>> call({
    required int id,
    required Object body,

    Dio? dio,
    Uri? uri,
    Timeout? timeout,
    CancelToken? cancelToken,
  }) async {
    final headers = {
      Headers.contentTypeHeader: setuContentType,
      SetuHeaders.id: id.toString(),
    };

    if (timeout != null) {
      headers[SetuHeaders.timeout] = timeout.toString();
    }

    final response = await (dio ?? RPC.dio).post(
      (uri ?? RPC.uri).toString(),
      data: body,
      cancelToken: cancelToken,
      options: Options(
        responseType: ResponseType.stream,
        sendTimeout: timeout?.duration,
        receiveTimeout: timeout?.duration,
        headers: {
          Headers.contentTypeHeader: setuContentType,
          SetuHeaders.id: id.toString(),
        },
      ),
    );

    final contentType = response.headers.value(Headers.contentTypeHeader);
    assertExpr(
      contentType == setuContentType,
      error: ProtocolError.new,
      message: () => 'unexpected content-type: ${contentType ?? "none"}',
    );

    return response.data.stream;
  }
}
